### PR TITLE
[Ready] Fixes >0.1u of chems running on_mob_add when they had been Qdeleted

### LINKED ISSUE #36152: You can gain a permanent speedboost by smoking
Round ID: 84467

If you add a small enough ratio of speed boosting chemicals to a cigarette/cigar/e-cig, and smoke it for a second, the speed boost you get will never run out.

[Reproduction]:
Prepare chemicals that boost your movement speed (meth, ephedrine, whatever else).
Water it down a bit. Let's say 20% of the boosting mix, 80% of anything else. 
Suck the mix in something you can smoke.
Smoke it.
You're now supersonic forever. You can take off the smoke and still keep it.
Bonus points by adding morphine, meaning you can run ultrafast no matter what you wear or how much damage you take.
So you can run around in a hardsuit and DUFFLEBAG at light speed.

You can even do this! [Sonic the clown](https://i.imgur.com/mXhyCOt.gif)
And yes it sounds as annoying as you think.

I only found this working for movement speed. Health regen, stun decrease or whatever else don't stick after smoking it. 
The health analyzer shows the reagents never entering you. At a higher ratio of the mix, it _does_, and then it just works normally.

### PR DESCRIPTION
**Pull request by Supermichael777**
banned for 43800 minutes edition

This time without hard locking on mob delete to on mob add
:cl:
fix: sub 0.1 unit doses of chemicals no longer add traits forever
code: Brave bull and tequila sunrise have checks to make sure that they actually added the effect they remove.  The other fix revealed this oversight
code: Updated chem's readme
/:cl:

Fixes: #36152

- [X] Test this locally

Its the cigarette bug but this time without touching a dozen files.  This could probably be done with a one line change but that would make acid cigs op according to the notes on the file. 

Previously on mob delete would run, then on mob add, then it would get deleted. this makes it so it only runs on mob delete then gets deleted in these cases.  

Adds safety checks to the people who made dumb assumptions.

### CODE REVIEW

**Comment by Kagus** [comment]
Ah, so is this what was causing the permanent stun/sleep immunity from smoking omega weed once? I always thought it was somehow associated with the brain trauma bath salts gives you.

Also, in the readme, "permenetnt" should probably be "permanent".

**Comment by XDTM** [comment]
It is indeed caused by the bath salts calling add and qdeling before remove, this should fix it

**Comment by Supermichael777** [comment]
Travis Broke

**Comment by Cyprex** [comment]
Let's fix chem processing Season 3 Episode 4

**Comment by Supermichael777** [comment]
Its round three with the let it sit for a week done, declare it stale and close it game, though the first two attempt were kinda kludged.
Fun fact closing a bad fix does not magically create a good fix and ~I haven't abuse it enough to get it fixed~ no one else seems to care.

**Comment by ExcessiveUseOfCobblestone** [comment]
so brave bull could have effectively made you invulnerable, surprised no one picked that bad boy up.

**Comment by XDTM** [comment]
yeah, it would stack, good catch

**Comment by Supermichael777** [comment]
brave bull actually couldn't because on mob delete ran FIRST, so it did nothing.

My first fix had a bug where it would run its old on mob delete. It set your max hp to -200. 
Admin heal would not fix it you had to manually vv

**Comment by Cyprex** [comment]
Thanks for fixing this, I'm surprised as well this has been left sitting around so long given the huge potential for exploits.

**Comment by Cyprex** [comment]
Thanks for fixing this, I'm surprised as well this has been left sitting around so long given the huge potential for exploits.

File: code/modules/reagents/chemistry/readme.md [hunk 1]
```diff
@@ -122,6 +122,19 @@ Reagents are all the things you can mix and fille in bottles etc. This can be an
 			reacts with a mob. The method var can be either
 			TOUCH or INGEST. You'll want to put stuff like
 			acid-facemelting in here.
+		
+		on_mob_add(mob/living/L)
+			Called when the reagent datum is created in the mobs holder.
+			Adding more while they still have some dosn't triger the effect a second time.
```
Comment (Kagus) [inline code/modules/reagents/chemistry/readme.md]: This line seems to be missing something, e.g. an E and a G.  Should probably be "doesn't trigger".

File: code/modules/reagents/chemistry/readme.md [hunk 2]
```diff
@@ -122,6 +122,19 @@ Reagents are all the things you can mix and fille in bottles etc. This can be an
...
+			has to be  at or above 0.1u to run
+			Happens last in the process
+		
+		on_mob_delete(mob/living/L)
+			Called when the reagent datum is deleted from
+			the mobs holder. Note that this does not check 
+			if on_mob_add has done anything. Saftey check
```
Comment (Kagus) [inline code/modules/reagents/chemistry/readme.md]: It's unsaf to spell "safety" as "saftey". Someone might get hur.

**Comment by Kagus** [comment]
"You no longer feel .x7"

Damn dude, you didn't have to add depression moodlets.

Also, thanks for finally fixing this. Even "just" the bath salts exploit was incredibly frustrating to deal with, I can't imagine what a stacked Alexander would look like.

**Comment by Supermichael777** [comment]
it looked like a man standing with a shield casually blocking an improbable level of stuff

**Comment by optimumtact** [comment]
Merging a bad fix doesn't make it a good one either

**Comment by Supermichael777** [comment]
Got a better fix

**Pull request closed without merging**